eews_backend/stream_processing/sandbox_predict.py

```diff
@@ -0,0 +1,170 @@
+#
+## Command in terminal docker (spark)
+# pip install influxdb_client motor keras tensorflow
+#
+## Notes
+# this code has not been run yet due to resource limitation on my end :(
+#
+###
+
+from influxdb_client import InfluxDBClient
+from motor import motor_asyncio
+
+from pyspark.sql import SparkSession
+from pyspark.sql.types import *
+import pyspark.sql.functions as f
+
+import datetime
+import pickle
+
+BOOTSTAP_SERVERS = "$"
+MONGO_URL = "$"
+MONGO_DATABASE = "$"
+INFLUXDB_ORG = "$"
+INFLUXDB_URL = "$"
+INFLUXDB_TOKEN = "$"
+
+model = pickle.load(open('model.pkl', 'rb'))
+
+def main():
+    spark, sc = init_spark()
+    model_broadcast = sc.broadcast(model)
+
+    df = spark \
+        .readStream \
+        .format("kafka") \
+        .option("kafka.bootstrap.servers", BOOTSTAP_SERVERS) \
+        .option("subscribe", "p-arrival") \
+        .load()
+    
+    schema = StructType([StructField('station', StringType()), StructField('time', StringType())])
+    schema_output = StructType([StructField('lat', StringType()), \
+                                StructField('long', StringType()), \
+                                StructField('depth', StringType()), \
+                                StructField('magnitude', StringType()), \
+                                StructField('time', StringType())])
+
+    df_json = df.selectExpr("CAST(value AS STRING) as json") \
+            .select(f.from_json('json', schema).alias('data')) \
+            .select('data.*')
+    
+    def apply_prediction(station, time):
+        import json
+        import datetime
+
+        # get data stasiun (mongo)
+        stations = get_nearest_station(station, 200000)
+
+        # get and preprocess data (influx)
+        seis_data = []
+
+        for station in stations:
+            data = read_seis_influx(station, datetime.datetime.strptime(time,'%Y-%m-%dT%H:%M:%S.%fZ')) # [[e,...,e],[n,...,n],[z,...,z]]
+            data_preprocessed = preprocess(data)
+            seis_data.append(data_preprocessed)
+
+        if len(stations) == 1:
+            seis_data *= 3
+        elif len(stations) == 2:
+            seis_data.append(seis_data[0])
+        
+        # predict
+        preds = model_broadcast.predict(seis_data)
+        data_mtr = denormalization(preds.iloc[0])
+        
+        return data_mtr
+
+    prediction_udf = f.udf(lambda data1, data2: apply_prediction(data1, data2), StringType())
+
+    query = df_json.select(f.from_json(prediction_udf(df_json.station, df_json.time), schema_output).alias('response'))\
+        .select('response.*') \
+        .writeStream \
+        .trigger(once=True) \
+        .format("console") \
+    
+    query.awaitTermination()
+
+
+def init_spark():
+    spark = SparkSession.builder.getOrCreate()
+    sc = spark.sparkContext
+    return spark, sc
+
+def read_seis_influx(station: str, time: datetime.datetime):
+    client = InfluxDBClient(
```

Comment:
Move all of the db client initialization to on top of the file so that it is only being initialized once. The current code means you will create a new connection to the db every time you want to query